resolver/resolver_test.go

```diff
@@ -417,27 +417,68 @@ func TestMultiError(t *testing.T) {
 	}
 }
 
-func TestTruncate(t *testing.T) {
-	tm := *newTruncated()
-	if !tm.Truncated {
+func TestTruncateSetTruncateBit(t *testing.T) {
+	testTruncateSetTruncateBit(t, dns.MinMsgSize)
+}
+
+func TestTruncateEdns0SetTruncateBit(t *testing.T) {
+	testTruncateSetTruncateBit(t, 4096)
+}
+
+func TestTruncateNoSetTruncateBit(t *testing.T) {
+	testTruncateNoSetTruncateBit(t, dns.MinMsgSize)
+}
+
+func TestTruncateEdns0NoSetTruncateBit(t *testing.T) {
+	testTruncateNoSetTruncateBit(t, 4096)
+}
+
+func testTruncateSetTruncateBit(t *testing.T, maxSize uint16) {
+	orig := *newMessage(maxSize)
+	tmp := orig
+	msg := *truncate(&tmp, true, true)
+	if !msg.Truncated {
 		t.Fatal("Message not truncated")
 	}
-	if l := tm.Len(); l > 512 {
+	if l := msg.Len(); l > int(maxSize) {
 		t.Fatalf("Message too large: %d bytes", l)
 	}
-	tm2 := *truncate(&tm, true)
-	if !tm2.Truncated {
+	// Replace any EDNS0 option that was stripped by the first
+	// truncate. This let's us test "truncate() of a large truncated message".
+	if opt := orig.IsEdns0(); opt != nil {
+		msg.SetEdns0(opt.UDPSize(), opt.Do())
+	}
+	msg2 := *truncate(&msg, true, true)
+	if !msg2.Truncated {
 		t.Fatal("Original truncation status was not preseved")
 	}
-	if tm2.Len() != tm.Len() {
+	if msg2.Len() != msg.Len() {
```

Comment:
All of the above notwithstanding - you're *correct* that this equality is a bug! :+1: